ENH: querying tool for the Planetary Ring Node

emolter opened the pull request:
The goal is to query the Planetary Ring Node's ephemeris tools at https://pds-rings.seti.org/tools/

pep8speaks commented:
Hello @emolter! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`astroquery/solarsystem/pds/__init__.py`](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/__init__.py):

> [Line 28:1](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/__init__.py#L28): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 29:33](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/__init__.py#L29): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 30:40](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/__init__.py#L30): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 31:40](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/__init__.py#L31): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 33:32](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/__init__.py#L33): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 34:40](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/__init__.py#L34): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 35:40](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/__init__.py#L35): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 37:32](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/__init__.py#L37): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 38:40](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/__init__.py#L38): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 39:40](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/__init__.py#L39): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 41:33](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/__init__.py#L41): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 42:40](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/__init__.py#L42): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 43:40](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/__init__.py#L43): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 46:1](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/__init__.py#L46): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 49:1](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/__init__.py#L49): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (3)
> [Line 51:1](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/__init__.py#L51): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 53:1](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/__init__.py#L53): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 55:14](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/__init__.py#L55): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file
> [Line 55:14](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/__init__.py#L55): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file
> [Line 55:14](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/__init__.py#L55): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file

* In the file [`astroquery/solarsystem/pds/core.py`](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py):

> [Line 25:1](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L25): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 27:1](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L27): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 29:1](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L29): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 32:1](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L32): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 40:1](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L40): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 47:1](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L47): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 49:1](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L49): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 50:1](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L50): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 51:5](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L51): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 54:1](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L54): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 66:75](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L66): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 67:1](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L67): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 68:30](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L68): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 70:48](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L70): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 74:1](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L74): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 76:1](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L76): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 81:1](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L81): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 91:1](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L91): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 93:9](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L93): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 94:1](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L94): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 100:1](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L100): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 105:67](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L105): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 107:25](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L107): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 109:32](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L109): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 117:42](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L117): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 118:47](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L118): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 119:24](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L119): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 119:29](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L119): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 120:25](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L120): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 120:30](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L120): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 121:23](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L121): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 121:28](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L121): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 122:24](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L122): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 122:29](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L122): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 123:21](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L123): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 123:66](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L123): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 124:21](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L124): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 124:35](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L124): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 125:24](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L125): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 125:29](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L125): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 126:29](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L126): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 127:25](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L127): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 128:26](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L128): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 129:21](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L129): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 130:22](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L130): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 131:23](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L131): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 138:1](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L138): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 146:1](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L146): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 149:67](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L149): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 151:1](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L151): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 157:1](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L157): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 159:1](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L159): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 164:1](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L164): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 166:1](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L166): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 167:1](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L167): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 168:1](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L168): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 169:9](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L169): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (3)
> [Line 170:1](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L170): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 171:46](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L171): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 171:48](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L171): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 175:1](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L175): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 181:1](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L181): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 189:9](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L189): [E722](https://duckduckgo.com/?q=pep8%20E722) do not use bare 'except'
> [Line 197:20](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L197): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 198:1](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L198): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 199:1](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L199): [E305](https://duckduckgo.com/?q=pep8%20E305) expected 2 blank lines after class or function definition, found 1
> [Line 199:27](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L199): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 202:1](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L202): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 206:1](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L206): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 206:1](https://github.com/astropy/astroquery/blob/92797bafc92759ecf8c4a3a97eb3167f69a8cb5e/astroquery/solarsystem/pds/core.py#L206): [W391](https://duckduckgo.com/?q=pep8%20W391) blank line at end of file

mkelley commented on astroquery/solarsystem/pds/core.py:
```diff
@@ -0,0 +1,447 @@
+    <https://pds-rings.seti.org/tools/>
+
+
+    # basic query for all six targets
```
Please place the example in a section:
```rst
Examples
--------
...
```

mkelley commented on astroquery/solarsystem/pds/core.py:
```diff
@@ -0,0 +1,447 @@
+        ----------
+        planet : str, required. one of Jupiter, Saturn, Uranus, or Neptune
+            Name, number, or designation of the object to be queried.
+        obs_time : str, in JD or MJD format. If no obs_time is provided, the current
```
Time should use an astropy `Time` object.  This will facilitate transformations between time formats and scales.

emolter replied:
is it ok/good to make it so that either an astropy Time object or a string can be passed? and then within the code if a string is passed then it becomes a Time object?

mkelley replied:
I think that is OK, but you will need to document the time scale assumed.

mkelley commented on astroquery/solarsystem/pds/core.py:
```diff
@@ -0,0 +1,447 @@
+
+    def __str__(self):
+        """
+        String representation of RingNodeClass object instance'
```
trailing single quote character

mkelley commented on astroquery/solarsystem/pds/core.py:
```diff
@@ -0,0 +1,447 @@
+
+        super().__init__()
+        self.planet = planet
+        self.obs_time = obs_time
```
Unless there is a reason to reuse planet and obs_time, or if you have a strong preference, I prefer to have everything happen in a single method call, e.g., `RingNode.ephemeris(planet, obs_time, observer, ...)`.

emolter replied:
ok, cool, I like this idea too. the only reason I did it the way I did was to look more like the jplhorizons module

mkelley commented on astroquery/solarsystem/pds/core.py:
```diff
@@ -0,0 +1,447 @@
+        >>> from astroquery.solarsystem.pds import RingNode
+        >>> uranus = RingNode(planet='Uranus',
+        ...                 obs_time='2017-01-01 00:00')
+        >>> eph = obj.ephemeris()  # doctest: +SKIP
```
By assigning `RingNode(...)` to `uranus`, this example becomes a little bit muddied.  I would think that `obs_time` is more closely aligned with the ephemeris() parameters, rather than the planet itself.

mkelley commented on astroquery/solarsystem/pds/core.py:
```diff
@@ -0,0 +1,447 @@
+        Parameters
+        ----------
+        self : RingNodeClass instance
+        observer_coords : three-element list/array/tuple of format (lat (deg), lon (deg east), altitude (m))
```
I recommend using astropy `Quantity` or `Angle` to take care of the units.  See `MPC.get_ephemeris()` where I also allowed `astropy.coordinates.EarthLocation`, but I'll leave it up to you if you want to implement that.

mkelley commented on astroquery/solarsystem/pds/core.py:
```diff
@@ -0,0 +1,447 @@
+                    "ephem",
+                    conf.planet_defaults[self.planet]["ephem"],
+                ),  # change hardcoding for other planets
+                ("time", self.obs_time),
```
What time scale is used here?  UTC?  TDB?

mkelley commented on astroquery/solarsystem/pds/core.py:
```diff
@@ -0,0 +1,447 @@
+                    "observatory",
+                    "Earth's center",
+                ),  # has no effect if viewpoint != observatory so can hardcode. no plans to implement calling observatories by name since ring node only names like 8 observatories
+                ("latitude", latitude),
```
If using `Angle` this, e.g., would be `latitude.degree`

mkelley commented on astroquery/solarsystem/pds/tests/test_ringnode.py:
```diff
@@ -0,0 +1,110 @@
+
+# Horizons has machinery here to mock request from a text file
+# is that necessary? why is that done?
+# wouldn't we want to know if the website we are querying changes something that makes code fail?
```
Remote tests are not routinely run.  With a mock test, we can run many tests in a short period of time without concern for server load or even internet connection.

emolter replied:
ok, I'll do my best with offline tests

mkelley reviewed (commented):
@emolter , this generally looks good.  Thanks for the contribution so-far.  I've looked at the top ~half of the code and have some suggestions.  I have not yet run it, but instead just providing some feedback on style at this time.

emolter commented:
@mkelley thanks very much for the helpful comments!  I had a few specific questions as well, if you don't mind.

1. the output of this code is three different groups of data: system-wide parameters `systemtable`, target-specific moon data `bodytable`, and target-specific ring data `ringtable`.  Tthe `bodytable` and `ringtable` make sense as astropy Tables: there are multiple moons (rows) that have the same parameters (columns) associated with them. But the `systemtable` is dict-like: it makes sense to be able to call, for example, `systemtable['opening_angle']` and return a single `Quantity` object.  My question is: what astropy data structure should I use for something dict-like?  And more broadly, is it okay to have three different tables returned from the query this way, or is there some higher-level data structure this stuff should go into?
2. what exactly are the testing requirements? is there any avenue for me to ask for help writing tests once the code operates in a good way?
3. I assume I need to use QTables instead of Tables when I have units - is that correct? sorry I'm so unfamiliar with Astropy generally...

Thanks again.  I'm going to work on this some more in the next week or so and see how far I can get

mkelley commented:
Regarding offline testing, there are some notes at: https://astroquery.readthedocs.io/en/latest/testing.html  Another benefit to offline tests is that the output from the "server" is always the same (i.e., the ephemeris never changes).  But definitely have a remote test or two so that we can address unexpected breaking format changes from the Node.

Regarding testing in general, you'll want to exercise and validate every line of code.  The output from codecov should help with that (it currently reads 16% covered), or you can run the tests offline and get a coverage report.  I would try using tox if you can, e.g.,:
```bash
tox -e py39-test-cov -- -Psolarsystem.pds --cov-report=html
```
Use py39 for python 3.9, or py38 for 3.8, etc.  This will only test the pds submodule.  Append `--remote-tests` to also run remote tests.  Then open .tmp/py38-test-cov/htmlcov/index.html and navigate to your submodule's files to see what was covered.  For more help, either send me a slack message or reply in this PR.

mkelley commented:
Regarding the output, I think three return variables is fine, and leaving `systemtable` as a dictionary is also OK.  As for `Table` vs. `QTable`, yeah, try out `QTable` first.

emolter commented:
@mkelley ok, I incorporated all of your comments - I put astropy units onto all the input and output quantities, including your suggestion to allow for an EarthLocation structure to be passed.  I'm now running into a few small problems: 

1. Do you have a suggested automatic de-linter?  The ones I'm using (autopep8, black) don't seem to fix everything, and sometimes make style changes that are incompatible with what tox is expecting (i.e., they seemingly make things worse!)
2. The offline` test_ephemeris_query` works on my machine (using the tox command you suggested above) but is failing to find the data files (tests/data/uranus_ephemeris.html) when the CI workflow is run.

Any advice on these two issues is much appreciated!

bsipocz commented:
> Do you have a suggested automatic de-linter? The ones I'm using (autopep8, black) don't seem to fix everything, and sometimes make style changes that are incompatible with what tox is expecting (i.e., they seemingly make things worse!)

please don't use black. autopep8 should work for most things, for the rest, I would suggest to set up flake to run in your editor, that will highlight potential pep8 issues while you develop the code.

mkelley commented:
Hi @emolter ,

Were you able to fix the mock tests?  The tests seem to be passing now.

mkelley commented on astroquery/solarsystem/pds/tests/test_ringnode.py:
```diff
@@ -0,0 +1,194 @@
+def test_bad_query_exception_throw():
+
+    with pytest.raises(ValueError):
+        pds.RingNode().ephemeris(planet="Mercury", obs_time="2022-05-03 00:00")
```
Here and elsewhere: `RingNode` is already an instance of `RingNodeClass`, so there is no need for the parentheses.  Just `pds.RingNode.ephemeris(...)` will suffice.

emolter replied:
oh ok, gotcha - now I see what you meant before.  will change

mkelley commented:
Also, the file coverage.xml seems to have been added by mistake.  Please remove it.

emolter commented:
Ok, I fixed everything you suggested @mkelley , thanks again for all your help!  I think this is nearly complete now.  I added some documentation and some more tests.  What is the next step at this stage?  What additional feedback do you have?

Could you please help me understand why the oldest dependencies and codecov/project checks are failing? The error messages are hard for me to parse.

bsipocz commented:
@emolter - I'll try to get back to this PR later this week for a round of review and suggestions for fixing the code to pass CI.

mkelley commented:
The "codecov/patch` details list all the lines that were not tested, e.g., astroquery/solarsystem/pds/core.py#L50 states that line 50 of that file was not covered by the tests.  I verified this locally and even with the remote tests enabled, the results were similar.  Most of the skipped lines are due to if-statement branching, e.g., lines 411-418 and 426-430 were not covered by the tests.  So, having Saturn and Neptune queries in the tests seems important.

emolter commented:
Ok, I added tests for the special cases for Saturn and Neptune.  It looks like now the oldest dependencies checks are failing because QTable doesn't like the "units" keyword.  But again all the tests pass for me with tox, even in python 3.7.  If I am correctly interpreting the error messages and this is indeed the problem, can you please advise me how to add units to a table that was read from ascii in a way that makes the oldest dependencies happy?

mkelley commented:
Perhaps you can annotate the table with a loop, adding the units to the `Column.unit` properties?
```python
>>> from astropy.table import Table, QTable
>>> tab = Table(rows=[{'a': 1}])
>>> tab
<Table length=1>
  a  
int64
-----
    1
>>> tab['a'].unit = 'km'
>>> tab
<Table length=1>
  a  
  km 
int64
-----
    1
>>> QTable(tab)
<QTable length=1>
   a   
   km  
float64
-------
    1.0
>>> QTable(tab)['a']
<Quantity [1.] km>
>>> import astropy
>>> astropy.__version__
'4.0'
>>> 
```

bsipocz commented on astroquery/solarsystem/pds/core.py:
```diff
@@ -0,0 +1,493 @@
+
+        return response
+
+    def _parse_ringnode(self, src):
```
I think it's cleaner if the full response is passed to this method rather than `response.text`, it will be slightly easier to debug

bsipocz commented on astroquery/solarsystem/pds/core.py:
```diff
@@ -0,0 +1,493 @@
+            # minor body table part 1
+            elif group.startswith("Body"):
+                group = "NAIF " + group  # fixing lack of header for NAIF ID
+                bodytable = ascii.read(
+                    group,
+                    format="fixed_width",
+                    col_starts=(0, 4, 18, 35, 54, 68, 80, 91),
+                    col_ends=(4, 18, 35, 54, 68, 80, 91, 102),
+                    names=(
+                        "NAIF ID",
+                        "Body",
+                        "RA",
+                        "Dec",
+                        "RA (deg)",
+                        "Dec (deg)",
+                        "dRA",
+                        "dDec",
+                    ),
+                )
+                units_list = [None, None, None, None, u.deg, u.deg, u.arcsec, u.arcsec]
+                bodytable = table.QTable(bodytable, units=units_list)
```
I know this is not consistent in astroquery, but you could initiatilize the table directly. I think it's much more legible, and also doesn't create another Table instance:

```
bodytable = QTable.read(group, format="ascii.fixed_width",
                        col_starts=(0, 4, 18, 35, 54, 68, 80, 91),
                        col_ends=(4, 18, 35, 54, 68, 80, 91, 102),
                        names=("NAIF ID", "Body", "RA", "Dec", "RA (deg)", "Dec (deg)", "dRA", "dDec"),
                        units=([None, None, None, None, u.deg, u.deg, u.arcsec, u.arcsec]))
```

bsipocz replied:
same logic applies to all other QTables or Tables below

emolter replied:
thank you for this! I was looking for a way to do this, and did not realize that you could pass ascii.fixed_width to the format keyword

bsipocz replied:
All the early modules parsed the ascii just as you did, so I suspect if we dig into the history there might have been some Table.read() limitations in those times. Basically with `Table.read()` you call the ready of the chosen format, and then have access all its options.

bsipocz commented on astroquery/solarsystem/pds/core.py:
```diff
@@ -0,0 +1,493 @@
+                )
+                units_list = [None, None, None, None, u.deg, u.deg, u.arcsec, u.arcsec]
+                bodytable = table.QTable(bodytable, units=units_list)
+                # for i in range(len(bodytable.colnames)):
```
remove commented code remnants

bsipocz commented on astroquery/solarsystem/pds/core.py:
```diff
@@ -0,0 +1,493 @@
+        bodytable = table.join(bodytable, bodytable2)  # concatenate minor body table
+        bodytable.add_index("Body")
+
+        systemtable["obs_time"] = Time(
+            obs_time, format="iso", scale="utc"
+        )  # add obs time to systemtable
```
again, keep this in one line (we let lines go up to 120 long, so no need to multi line a one liner simple expression)

bsipocz commented on astroquery/solarsystem/pds/core.py:
```diff
@@ -0,0 +1,493 @@
+
+        return systemtable, bodytable, ringtable
+
+    def _parse_result(self, response, verbose=None):
```
Why not have the content of `_parse_ringnode` in here, why the need for two parse methods?

bsipocz commented on astroquery/solarsystem/pds/core.py:
```diff
@@ -0,0 +1,493 @@
+        self.last_response = response
+        try:
+            systemtable, bodytable, ringtable = self._parse_ringnode(response.text)
+        except Exception as ex:
```
again, no need for the exception. If you expect some errors from the server, always can add a `response.raise_for_status()` after calling `self._request`

bsipocz commented on astroquery/solarsystem/pds/tests/test_ringnode.py:
```diff
@@ -0,0 +1,328 @@
+import pytest
+import os
+from collections import OrderedDict
```
everything is really an ordereddict these days, so no need to be specific.

bsipocz commented on docs/solarsystem/pds/pds.rst:
```diff
@@ -0,0 +1,132 @@
+.. doctest-skip-all
```
don't skip tests in the documentation, but use the `doctest-remote-data` directive to flag code snippets that are expected to reach the internet

bsipocz commented on docs/solarsystem/pds/pds.rst:
```diff
@@ -0,0 +1,132 @@
+   >>> from astroquery.solarsystem.pds import RingNode
+   >>> systemtable, bodytable, ringtable = RingNode.ephemeris(planet='Uranus',
+   ...                 obs_time='2024-05-08 22:39',
+   ...                 location = (-23.029 * u.deg, -67.755 * u.deg, 5000 * u.m))  # doctest: +SKIP
```
don't skip

bsipocz commented on docs/solarsystem/pds/pds.rst:
````diff
@@ -0,0 +1,132 @@
+
+``planet`` must be one of ['mars', 'jupiter', 'uranus', 'saturn', 'neptune', 'pluto'] (case-insensitive)
+
+``obs_time`` is the datetime to query. Accepts a string in format 'YYYY-MM-DD HH:MM' (UTC assumed), or an ```astropy.Time`` object. If no obs_time is provided, the current time is used.
````
use sphinx links, add them in single backticks with their correct namespace
```suggestion
``obs_time`` is the datetime to query. Accepts a string in format 'YYYY-MM-DD HH:MM' (UTC assumed), or an `~astropy.time.Time` object. If no obs_time is provided, the current time is used.
```

bsipocz replied:
This is a comment for here and below, too. Keep the double backticks for the parameter names, and anything else that you want to code highlight, but every public API should be linked.

bsipocz commented on astroquery/solarsystem/pds/core.py:
```diff
@@ -0,0 +1,493 @@
+            systemtable, bodytable, ringtable = self._parse_ringnode(response.text)
+        except Exception as ex:
+            try:
+                self._last_query.remove_cache_file(self.cache_location)
```
Why do you do this? To disable caching remove the kwarg instead and pass cache=False to `self._request`.

emolter replied:
honestly I have no idea how to do this "right," I pretty much copied everything here from the JPL Horizons query.  May I request some help here? I'm happy with however this ends up

bsipocz replied:
OK, let's come back to this once you push the updates. Without fixing it, at least in the mocks, the test will fail.

mkelley replied:
And I was following the SIMBAD example 😄 😵‍💫 I'll look at a few updates to jplhorizons now.

bsipocz reviewed (commented):
A lot of comments, many of them are nitpicks, some are for code clean-ups.
I think is very close to be being ready, but could benefit some simplification.

bsipocz commented on astroquery/solarsystem/pds/core.py:
```diff
@@ -0,0 +1,493 @@
+    def ephemeris_async(
+        self,
+        planet,
+        obs_time=None,
```
https://peps.python.org/pep-3102/

```
    def ephemeris_async(self, planet, *, obs_time=None, location=None, neptune_arcmodel=3, 
                        get_query_payload=False, get_raw_response=False, cache=True):

```

emolter commented on astroquery/solarsystem/pds/core.py:
```diff
@@ -0,0 +1,493 @@
+
+        if obs_time is None:
+            obs_time = Time.now().strftime("%Y-%m-%d %H:%M")
+            warnings.warn("obs_time not set. using current time instead.")
```
I haven't really worked with logs before - can you point me to an example of how this is done?  I have just removed the line for now

bsipocz replied:
There are a couple of modules using the logger already, e.g. look for `log.warning` in the codebase (e.g. the main `query.py` file)

bsipocz commented on astroquery/solarsystem/pds/core.py:
```diff
@@ -0,0 +1,493 @@
+            Epsilon    298.022            0.0
+        """
+
+        URL = conf.pds_server
```
@emolter - yes, moving it to the initialization of the class is what I meant, the other comments are bringing in a parallel discussions to about how it's best done to provide a way for users to override config items for class instances. 
Anyway, hard-wiring it in the methods like this is very much unrecommended.

emolter commented on astroquery/solarsystem/pds/core.py:
```diff
@@ -0,0 +1,493 @@
+        response = self._request(
+            "GET", URL, params=request_payload, timeout=self.TIMEOUT, cache=cache
+        )
+        self.uri = response.url
```
I put this in for consistency with the JPLHorizons ephemeris query, but I'm happy to delete it if desired

mkelley replied:
For jplhorizons, since we are using HTTP GET, exposing the URL as a property for the user has helped with debugging code and user errors.

emolter commented on astroquery/solarsystem/pds/core.py:
```diff
@@ -0,0 +1,493 @@
+        ringtable : `astropy.Table`
+        """
+
+        self.raw_response = src
```
I put this in for consistency with the JPLHorizons ephemeris query, but I'm happy to delete it if desired

bsipocz commented on astroquery/solarsystem/pds/core.py:
```diff
@@ -0,0 +1,493 @@
+
+        # set return_raw flag, if raw response desired
+        if get_raw_response:
+            self.return_raw = True
```
OK, fair enough. I believe this was then an oversight that we didn't catch at the time of the jpl review. In the standard astroquery API, the "raw" response is available via the async methods, maybe @mkelley can shed a light on it why this and uri was added as an extra in jplhorizons.

mkelley replied:
I didn't write jplhorizons, so I don't know why, but I can edit it now to avoid propagating things like this.

emolter reviewed (commented):
Thanks very much for the reviews @bsipocz @mkelley @eerovaher !  I went through all of the comments and made as many changes as I could.  I'm about to commit those changes so you can look at what I did.  I have some outstanding questions - some stuff went over my head given that this is my first ever pull request.  I added comments inside the review as well, but here are the high-level things:

1.  is there a way for me to build and view the documentation locally, so I can see if all the linking works properly?
2. It sounded like there was some debate over how exactly to do the request handling.  Would someone be willing to just make the edits they see fit?
3. Related, @bsipocz suggested removing self.uri and the return_raw flag, among other things. I don't really understand the benefits/drawbacks of doing this - TBH I just copy-pasted what was in the Horizons query tool.  Can someone please advise?

bsipocz commented:
> is there a way for me to build and view the documentation locally, so I can see if all the linking works properly?

`python setup.py build_sphinx` will build the docs. It should come back without any warnings

bsipocz commented:
> Related, @bsipocz suggested removing self.uri and the return_raw flag, among other things. I don't really understand the benefits/drawbacks of doing this - TBH I just copy-pasted what was in the Horizons query tool. Can someone please advise?

imo they should be removed from horizons, too. As I mentioned it's likely slipped through the original code review. Basically all that info is available with the async method,  no need to duplicate and store it on the instance.

bsipocz commented:
> It sounded like there was some debate over how exactly to do the request handling. Would someone be willing to just make the edits they see fit?

I lost track what the debate was, could you cross link. If this is the only thing standing I can fit it up before merging.

bsipocz commented:
Ahh, OK, for the second one, let's see what the failure will say, we may need to patch your MockResponse in the test, but leave the code as is, until we have a fixed cache mechanism. That part as I now recall was to workaround some failures users came by picking up outdated cached queries for jplhorizons.

emolter commented:
Hooray, this is passing all the tests!  Let me know if you see anything else I should change.  

I still didn't know what to do with https://github.com/astropy/astroquery/pull/2358#discussion_r880860547 so that might still need looking at

emolter commented:
Is there a back-compatible way to rewrite `epoch.utc.to_value("iso", subfmt="date_hm"))`?  This passes the tests with tox on my machine but fails here.

mkelley commented:
How about parsing the string?  `Time.now().utc.iso[:16]`

eerovaher commented on astroquery/solarsystem/pds/__init__.py:
```diff
@@ -0,0 +1,82 @@
+    Configuration parameters for `astroquery.solarsystem.pds`.
+    """
+
+    # server settings
+    pds_server = _config.ConfigItem(
+        ["https://pds-rings.seti.org/cgi-bin/tools/viewer3_xxx.pl?"], "Ring Node"
+    )
```
If `ConfigItem` is created using a `list` (like is done here) then that `list` contains all allowed values, unless the `cfgtype='list'` is explicitly used, see https://docs.astropy.org/en/stable/config/index.html#item-types-and-validation for more details.

This configuration item has only one allowed value, which means it is not configurable at all and should not be managed by the configuration system.

eerovaher commented on astroquery/solarsystem/pds/core.py:
```diff
@@ -0,0 +1,368 @@
+# 1. standard library imports
+
+from collections import OrderedDict
```
If you are not reordering any dictionaries frequently then you should not be using `OrderedDict`.
```suggestion
```

eerovaher commented on astroquery/solarsystem/pds/core.py:
```diff
@@ -0,0 +1,368 @@
+    <https://pds-rings.seti.org/tools/>
+    """
+
+    TIMEOUT = conf.timeout
```
Use the API proposed in #2341 for the timeout. See also #544 and #2291.

eerovaher commented on astroquery/solarsystem/pds/core.py:
```diff
@@ -0,0 +1,368 @@
+        self.URL = conf.pds_server
+        self.planet_defaults = conf.planet_defaults
+
+    def __str__(self):
+        """
+        String representation of `~RingNodeClass` object instance
+
+        Examples
+        --------
+        >>> from astroquery.solarsystem.pds import RingNode
+        >>> nodeobj = RingNode
+        >>> print(nodeobj)  # doctest: +SKIP
+        PDSRingNode instance
+        """
+        return "PDSRingNode instance"
```
What the `__str__()` method does would be obvious even without the docstring.

eerovaher commented on astroquery/solarsystem/pds/core.py:
```diff
@@ -0,0 +1,368 @@
+            epoch = Time(epoch, format='jd')
+        elif isinstance(epoch, str):
+            epoch = Time(epoch, format='iso')
+        elif epoch is None:
+            epoch = Time.now()
+            log.warning("Observation time not set. Using current time.")
```
Using the current time if `epoch=None` is what the API says it does, so there is no need for a warning. But even if you do want to emit one you should use the Python warnings system instead of the logging system. The warnings from the logging system are meant for situations which the client application can do nothing about, but the function arguments are very much under the user's control. See also https://docs.python.org/3/howto/logging.html#when-to-use-logging

emolter replied:
I had implemented this using Python's warnings, but @bsipocz thought it was better as a logged warning... I guess I'm happy just deleting it altogether

eerovaher commented on astroquery/solarsystem/pds/core.py:
```diff
@@ -0,0 +1,368 @@
+            viewpoint = "latlon"
+            if isinstance(location, EarthLocation):
+                loc = location.geodetic
+                longitude = loc[0].deg
+                latitude = loc[1].deg
+                altitude = loc[2].to(u.m).value
+            elif hasattr(location, "__iter__"):
+                latitude = Angle(location[0]).deg
+                longitude = Angle(location[1]).deg
+                altitude = u.Quantity(location[2]).to("m").value
```
I would find it less confusing if your code would expect longitude and latitude in the same order as `EarthLocation`.

eerovaher commented on astroquery/solarsystem/pds/tests/test_ringnode.py:
```diff
@@ -0,0 +1,303 @@
+        location=(10.0 * u.deg, -120.355 * u.deg, 1000 * u.m),
+    )
+    # check system table
+    assert_quantity_allclose(
+        [
+            -56.12233,
+            -56.13586,
+            -56.13586,
+            -56.01577,
+            0.10924,
+            354.11072,
+            354.12204,
+            2947896667.0,
+            3098568884.0,
+            10335.713263,
+        ],
+        [
+            systemtable["sub_sun_lat"].to(u.deg).value,
+            systemtable["sub_sun_lat_min"].to(u.deg).value,
+            systemtable["sub_sun_lat_max"].to(u.deg).value,
+            systemtable["opening_angle"].to(u.deg).value,
+            systemtable["phase_angle"].to(u.deg).value,
+            systemtable["sub_sun_lon"].to(u.deg).value,
+            systemtable["sub_obs_lon"].to(u.deg).value,
+            systemtable["d_sun"].to(u.km).value,
+            systemtable["d_obs"].to(u.km).value,
+            systemtable["light_time"].to(u.second).value,
+        ],
```
The point of using `assert_quantity_allclose()` is to compare both the values and the units. If you are comparing the numerical values only then you would be better off using `np.allclose()`.

eerovaher commented on astroquery/solarsystem/pds/core.py:
```diff
@@ -0,0 +1,368 @@
+
+        systemtable["epoch"] = Time(epoch, format="iso", scale="utc")  # add obs time to systemtable
+
+        return systemtable, bodytable, ringtable
```
Would it be a better idea to make the information in `systemtable` available in the metadata of `bodytable` and `ringtable` instead of returning it separately?

mkelley replied:
That's a good idea.

eerovaher commented on astroquery/solarsystem/pds/__init__.py:
```diff
@@ -0,0 +1,82 @@
+    Configuration parameters for `astroquery.solarsystem.pds`.
+    """
+
+    # server settings
+    pds_server = _config.ConfigItem(
+        "https://pds-rings.seti.org/cgi-bin/tools/viewer3_xxx.pl?", "Ring Node"
+    )
```
Does this service have any mirrors? If not then providing a configuration item for the server might not be worth the effort.

eerovaher commented on astroquery/solarsystem/pds/core.py:
```diff
@@ -0,0 +1,354 @@
+        Instantiate Planetary Ring Node query
+        '''
+        super().__init__()
+        self.url = conf.pds_server
+        self.timeout = conf.timeout
+        self.planet_defaults = conf.planet_defaults
```
This will prevent changing configuration items at runtime from having any effect. The correct thing to do would be to follow the template in #2341. I have added a section to the opening post there that explains why that is the best solution.

eerovaher commented on astroquery/solarsystem/pds/core.py:
```diff
@@ -0,0 +1,354 @@
+        self.last_response = response
+        try:
+            self._last_query.remove_cache_file(self.cache_location)
+        except (FileNotFoundError, AttributeError):
```
It looks like this `AttributeError` case is hiding some actual problem that should be fixed instead.

eerovaher commented on docs/solarsystem/pds/pds.rst:
```diff
@@ -0,0 +1,116 @@
+
+   >>> bodytable, ringtable = RingNode.ephemeris(planet='Venus',
+   ...                 epoch='2024-05-08 22:39',
+   ...                 location = (-67.755 * u.deg, -23.029 * u.deg, 5000 * u.m)) # doctest: +IGNORE_EXCEPTION_DETAIL
```
Is `+IGNORE_EXCEPTION_DETAIL` needed here? I would have thought the ellipsis in the traceback message should have been sufficient.

eerovaher commented on docs/solarsystem/pds/pds.rst:
```diff
@@ -0,0 +1,106 @@
+        Lambda        0.0            0.0
+       Epsilon    298.022            0.0
+
+``planet`` must be one of ['mars', 'jupiter', 'uranus', 'saturn', 'neptune', 'pluto'] (case-insensitive)
```
Now that there is an explicit example of an invalid planet query it might be possible to rephrase this sentence.

eerovaher commented on astroquery/solarsystem/pds/core.py:
```diff
@@ -0,0 +1,406 @@
+        ringtable : `astropy.QTable`
+        """
+        self.last_response = response
+        try:
+            self._last_query.remove_cache_file(self.cache_location)
+        except FileNotFoundError:
+            # this is allowed: if `cache` was set to False, this
+            # won't be needed
+            pass
```
You have taken this snippet of code from the `jplhorizons` sub-package, but you haven't copied the following test code:
 https://github.com/astropy/astroquery/blob/be0938f9bd9a989e25cf98d255635278ad883633/astroquery/jplhorizons/tests/test_jplhorizons.py#L62-L66 This must be the cause of the unexpected `AttributeError` failures in the tests.

emolter replied:
ok, I added this test.  it doesn't seem to solve the problem, though.  I'm not sure how to diagnose this especially because all the tests pass for me. any ideas?  Thanks again for all your help with this, by the way

emolter replied:
how would I modify `patch_request` to set the `_last_query`? 

that does seem like the way to go because otherwise I would no longer be able to use the `pds.RingNode.ephemeris()` syntax and instead would need to instantiate a `RingNode` object, then set `_last_query`, and finally call `ephemeris()`.  Yes, I am running tests with tox, which is why it's so confusing that it works on my machine... maybe I need to use different tox options.  BTW, I have been trying to model all the monkeypatch stuff off of what the jplhorizons astroquery tool did, and I'm not sure what I am missing.

bsipocz replied:
I see these failures when running only the local tests. When the remotes are also run all tests pass, showing that something is off with the mocking. I'm investigating of how it's all fine for jplhorizons but going off here.

bsipocz commented on astroquery/solarsystem/pds/core.py:
```diff
@@ -0,0 +1,406 @@
+    def _timeout(self):
+        return conf.timeout if self.timeout is None else self.timeout
+
+    def __str__(self):
```
This maybe a bit superfluous

bsipocz commented on astroquery/solarsystem/pds/core.py:
```diff
@@ -0,0 +1,406 @@
+        return "PDSRingNode instance"
+
+    def ephemeris_async(self, planet, *, epoch=None, location=None, neptune_arcmodel=3,
+                            get_query_payload=False, get_raw_response=False, cache=True):
```
remove `get_raw_response` from the signature. Also add `cache` to the docstring below

bsipocz commented on astroquery/solarsystem/pds/core.py:
```diff
@@ -0,0 +1,406 @@
+        bodytable : `astropy.QTable`
+        ringtable : `astropy.QTable`
+        """
+        self.last_response = response
```
do we really need to store the last response?

bsipocz commented on docs/solarsystem/pds/pds.rst:
```diff
@@ -0,0 +1,116 @@
+``RingNode`` object is instantiated and the :meth:`~astroquery.solarsystem.pds.RingNodeClass.ephemeris` method is called. The following example queries the
+ephemerides of the rings and small moons around Uranus as viewed from ALMA:
+
+.. code-block:: python
```
you need to do all these snippets in a `doctest-remote-data` context, so they are skipped for local only tests. The context need to be added for each and every snippet (we don't yet support adding only one for the top of the file).

bsipocz reviewed (commented):
minor comments only, the one major remaining thing is to figure out the issue with the mocks and fix it, I'll try to get that figured out independent of the rest of my comments here.

bsipocz commented:
One more nitpick: please `git mv` the test files from ringnode to be `pds`.

bsipocz commented:
OK, so the big difference compared to the jplhorizons example is that they don't hit the part of the code where `_last_query` is being used, so it's enough for that module to patch only one of the test functions.

There are two ways to tackle this: 
 - use an explicit fresh instance in each tests. you basically do this, but don't patch its `last_query` with the dummy. So your typical test function will start with the following, and then run the methods of this instance.
 ```
pds_inst = pds.RingNode()
pds_inst._last_query = AstroQuery('GET', 'http://dummy')
```

 - create a patched instance and use it as a fixture. e.g. alma does this for the remote tests

emolter commented:
@bsipocz thanks for your continued help with this!  I finally made some more progress - sorry for the delay.  But now the "oldest dependencies" checks are failing due to an error I don't understand, something to do with how I'm instantiating the astropy tables... would you please be able to shed some light on what's happening there? once again, this succeeds on my machine using tox and py3.7

bsipocz commented on astroquery/solarsystem/pds/core.py:
```diff
@@ -0,0 +1,422 @@
+
+            # input parameters. only thing needed is epoch
+            if group.startswith("Observation"):
+                epoch = group.split("\n")[0].split("e: ")[-1].strip(", \n")
```
without looking to much into the details, I think this line (and the previous ones) mess up the Windows tests, namely that your line-ending assumptions may not be true for windows.

emolter replied:
Thanks for looking into it @bsipocz, and yes, I see that the error comes from those lines.  I am pretty lost though as to how to play with or debug the code on a Windows machine.  In fact, it seems to only throw an error for specific Windows machines, since an undergrad I advise is working with the code on Windows and is not experiencing this issue.

eerovaher replied:
Have you tried using [`str.splitlines()`](https://docs.python.org/3/library/stdtypes.html#str.splitlines) instead of `str.split("\n")`?

emolter replied:
thanks for the suggestion!  I will look into it tomorrow.

emolter replied:
a version of this suggestion worked! thank you so much @eerovaher !  I *think* this is ready now

bsipocz replied:
I don't see the modified version with splitlines, could you please simplify the manual line ending checks from above and use the method as it was suggested?

emolter commented:
@eerovaher @mkelley @bsipocz Thanks again for your help with this piece of code!  Is it still possible to get this merged for 5.2? if so, what is needed from me to make that happen?

eerovaher commented:
`astroquery` release schedule is independent from `astropy`.

bsipocz commented:
OK, so first of all, I go ahead and rebase this to pick up more recent changes of CI, then I'll do a final review.

bsipocz commented on docs/solarsystem/pds/pds.rst:
```diff
@@ -0,0 +1,116 @@
+    :no-inheritance-diagram:
+
+.. _NASA Planetary Data System's Ring Node System: https://pds-rings.seti.org/
+.. _astropy Quantity: https://docs.astropy.org/en/stable/units/quantity.html
+.. _astropy table: http://docs.astropy.org/en/stable/table/index.html
+.. _astropy units: http://docs.astropy.org/en/stable/units/index.html
```
No need for these links here, intersphinx is used above.

bsipocz commented on docs/solarsystem/pds/pds.rst:
```diff
@@ -0,0 +1,116 @@
+	dict_keys(['sub_sun_lat', 'sub_sun_lat_min', 'sub_sun_lat_max', 'opening_angle', 'phase_angle', 'sub_sun_lon', 'sub_obs_lon', 'd_sun', 'd_obs', 'light_time', 'epoch'])
+
+
+Acknowledgements
```
We don't normally do an acknowledgement section, but start with the info of what the modules uses at the top.

bsipocz commented on docs/solarsystem/pds/pds.rst:
```diff
@@ -0,0 +1,116 @@
+``RingNode`` object is instantiated and the :meth:`~astroquery.solarsystem.pds.RingNodeClass.ephemeris` method is called. The following example queries the
+ephemerides of the rings and small moons around Uranus as viewed from ALMA:
+
+.. code-block:: python 
+
+   >>> from astroquery.solarsystem.pds import RingNode
+   >>> import astropy.units as u
+   >>> bodytable, ringtable = RingNode.ephemeris(planet='Uranus',
+   ...                 epoch='2024-05-08 22:39',
+   ...                 location = (-67.755 * u.deg, -23.029 * u.deg, 5000 * u.m)) # doctest: +REMOTE_DATA
+   >>> print(ringtable) # doctest: +REMOTE_DATA
```
For multi-line tests that need remote data, use the directive (
`.. doctest-remote-data::`), here and below (where I don't repeat the suggestion):
```suggestion
.. doctest-remote-data:: 

   >>> from astroquery.solarsystem.pds import RingNode
   >>> import astropy.units as u
   >>> bodytable, ringtable = RingNode.ephemeris(planet='Uranus',
   ...                 epoch='2024-05-08 22:39',
   ...                 location = (-67.755 * u.deg, -23.029 * u.deg, 5000 * u.m))
   >>> print(ringtable)
```

bsipocz commented on astroquery/solarsystem/pds/tests/test_pds_remote.py:
```diff
@@ -0,0 +1,82 @@
+        )
+        # check system table
+        systemtable = bodytable.meta
+        assert np.allclose(
+            [
+                -56.12233,
+                -56.13586,
+                -56.13586,
+                -56.01577,
+                0.10924,
+                354.11072,
+                354.12204,
+                2947896667.0,
+                3098568884.0,
+                10335.713263,
+            ],
+            [
+                systemtable["sub_sun_lat"].to(u.deg).value,
+                systemtable["sub_sun_lat_min"].to(u.deg).value,
+                systemtable["sub_sun_lat_max"].to(u.deg).value,
+                systemtable["opening_angle"].to(u.deg).value,
+                systemtable["phase_angle"].to(u.deg).value,
+                systemtable["sub_sun_lon"].to(u.deg).value,
+                systemtable["sub_obs_lon"].to(u.deg).value,
+                systemtable["d_sun"].to(u.km).value,
+                systemtable["d_obs"].to(u.km).value,
+                systemtable["light_time"].to(u.second).value,
+            ],
+            rtol=1e-2,
+        )
```
Here and below, too:

please refrain from using black for astroquery, and condense these 25 lines into something sensible, all of these should fit into ~5 lines.

bsipocz commented on astroquery/solarsystem/pds/tests/test_pds.py:
```diff
@@ -0,0 +1,327 @@
+        epoch="2021-10-07 07:25",
+    )
+
+    assert np.allclose(
+                    [249.23097,
+                     250.34081
+                     ],
+                     [
+                     ringtable[ringtable.loc_indices["F"]]["pericenter"].to(u.deg).value,
+                     ringtable[ringtable.loc_indices["F"]]["ascending node"].to(u.deg).value
+                     ], rtol=1e-3)
```
please do reformat

bsipocz commented on astroquery/solarsystem/pds/tests/test_pds.py:
```diff
@@ -0,0 +1,327 @@
+        neptune_arcmodel=2
+    )
+
+    assert np.allclose(
+                    [63.81977,
+                     55.01978,
+                     44.21976,
+                     40.41978,
+                     26.41978,
+                     64.81977,
+                     59.11976,
+                     45.21976,
+                     43.41978,
+                     36.01978],
+                     [ringtable[ringtable.loc_indices["Courage"]]["min_angle"].to(u.deg).value,
+                     ringtable[ringtable.loc_indices["Liberte"]]["min_angle"].to(u.deg).value,
+                     ringtable[ringtable.loc_indices["Egalite A"]]["min_angle"].to(u.deg).value,
+                     ringtable[ringtable.loc_indices["Egalite B"]]["min_angle"].to(u.deg).value,
+                     ringtable[ringtable.loc_indices["Fraternite"]]["min_angle"].to(u.deg).value,
+                     ringtable[ringtable.loc_indices["Courage"]]["max_angle"].to(u.deg).value,
+                     ringtable[ringtable.loc_indices["Liberte"]]["max_angle"].to(u.deg).value,
+                     ringtable[ringtable.loc_indices["Egalite A"]]["max_angle"].to(u.deg).value,
+                     ringtable[ringtable.loc_indices["Egalite B"]]["max_angle"].to(u.deg).value,
+                     ringtable[ringtable.loc_indices["Fraternite"]]["max_angle"].to(u.deg).value,
+                      ], rtol=1e-3)
```
please do reformat

bsipocz commented on astroquery/solarsystem/pds/tests/test_pds.py:
```diff
@@ -0,0 +1,327 @@
+    )
+    # check system table
+    systemtable = bodytable.meta
+    assert np.allclose(
+        [
+            -56.12233,
+            -56.13586,
+            -56.13586,
+            -56.01577,
+            0.10924,
+            354.11072,
+            354.12204,
+            2947896667.0,
+            3098568884.0,
+            10335.713263,
+        ],
+        [
+            systemtable["sub_sun_lat"].to(u.deg).value,
+            systemtable["sub_sun_lat_min"].to(u.deg).value,
+            systemtable["sub_sun_lat_max"].to(u.deg).value,
+            systemtable["opening_angle"].to(u.deg).value,
+            systemtable["phase_angle"].to(u.deg).value,
+            systemtable["sub_sun_lon"].to(u.deg).value,
+            systemtable["sub_obs_lon"].to(u.deg).value,
+            systemtable["d_sun"].to(u.km).value,
+            systemtable["d_obs"].to(u.km).value,
+            systemtable["light_time"].to(u.second).value,
+        ],
+        rtol=1e-2,
```
please do reformat, this and all array in this test. We don't use black here, and these examples highlight one of the reasons, namely that it cannot handle arrays.

bsipocz commented on astroquery/solarsystem/pds/core.py:
```diff
@@ -0,0 +1,434 @@
+        # and make system-wide parameters metadata of bodytable and ringtable
+        # FAILING
+        # UnboundLocalError: local variable 'epoch' referenced before assignment
+        # only on Windows (wtf?)
```
please do remove comments like this.

bsipocz reviewed (commented):
This is now reasonably close for getting merged. It still requires a changelog entry, and a bit of cleanup, please see the comments.

Also, if you could squash the commits into logical chunks it would be appreciated (otoh, I'm happy to do that prior to merging)

bsipocz commented:
@emolter - thanks for following up on the review. I have have a detailed look on the remaining windows issues later this evening and eventually get this merged!

emolter commented:
Ok @bsipocz, thanks again so much for your helpful comments on this piece of code.  I removed all explicit calls to \n, preferring splitlines(), changing split(' \n') to simply split(), and when absolutely necessary using os.linesep.  I changed all the other things you suggested; at least, I think I did.  It's now passing again.

Sorry if these are basic questions, but (1) how/where do I write a changelog entry and what is required, and (2) how do I squash commits into logical chunks?

bsipocz commented:
- changelog is a 1-2 sentence in `CHANGES.rst`, right at the top under the new tools and services, something like this:
```
solarsystems.pds
^^^^^^^^^^^^^

- New module to access the Planetary Data System's Ring Node System. [#2358]
```


Squashing is done as an "interactive rebase" where you can select commits to get merged together. Mostly to remove the back-and-forth experiments with fixing the tests, etc. I'm totally happy to deal with this though.

emolter commented:
I just committed the changelog entry, and I will leave the chunking of commits to you.  Thanks again, and let me know what else needs to be done.

eerovaher commented on astroquery/solarsystem/pds/core.py:
```diff
@@ -0,0 +1,427 @@
+                loc = location.geodetic
+                longitude = loc[0].deg
+                latitude = loc[1].deg
+                altitude = loc[2].to(u.m).value
```
`.to_value(unit)` is faster than `.to(unit).value`.

eerovaher commented on astroquery/solarsystem/pds/core.py:
```diff
@@ -0,0 +1,427 @@
+        """
+        try:
+            self._last_query.remove_cache_file(self.cache_location)
+        except (FileNotFoundError, OSError):
```
`FileNotFoundError` is a subclass of `OSError`, so I don't see why it would be necessary to check for both.

eerovaher commented on astroquery/solarsystem/pds/core.py:
```diff
@@ -0,0 +1,427 @@
+
+            # input parameters. only thing needed is epoch
+            if group.startswith("Observation"):
+                epoch = group.splitlines()[0].split("e: ")[-1].strip()
```
There is a method for splitting from the right:
```suggestion
                epoch = group.splitlines()[0].rsplit("e: ", 1)[1].strip()
```

eerovaher commented on astroquery/solarsystem/pds/core.py:
```diff
@@ -0,0 +1,427 @@
+                                        col_starts=(0, 4, 18, 35, 54, 68, 80, 91),
+                                        col_ends=(4, 18, 35, 54, 68, 80, 91, 102),
+                                        names=bodytable_names)
+                                        # units=(bodytable_units)) # this much cleaner way of adding units is supported in later versions but not in 3.7
```
Are you sure about this? I would find it very strange if the Python version would have any effect on the keyword arguments `QTable.read()` can accept.

bsipocz replied:
The comment is wrong in the sense that it's not python 3.7 related, but oldest supported astropy related as the `units` kwarg got added in 4.1. 

We haven't really formalized the dropping of versions, but I kind of like to follow NEP29/SPEC0 with the caveat that for astroquery we do the drops more on need bases rather than automatically. Anyway, 24 months have passed since 4.1, so let's drop 4.0 support outside of this PR.

eerovaher commented on astroquery/solarsystem/pds/core.py:
```diff
@@ -0,0 +1,427 @@
+
+            # ring plane data
+            elif group.startswith("Ring s"):
+                lines = group.splitlines()
+                for line in lines:
```
```suggestion
                for line in group.splitlines():
```

eerovaher commented on astroquery/solarsystem/pds/core.py:
```diff
@@ -0,0 +1,427 @@
+                for line in lines:
+                    l = line.split(":")
+                    if "Ring sub-solar latitude" in l[0]:
+                        [sub_sun_lat, sub_sun_lat_min, sub_sun_lat_max] = [
```
```suggestion
                        sub_sun_lat, sub_sun_lat_min, sub_sun_lat_max = [
```

eerovaher commented on astroquery/solarsystem/pds/core.py:
```diff
@@ -0,0 +1,427 @@
+                        )
+                    elif "Sub-observer longitude" in l[0]:
+                        systemtable["sub_obs_lon"] = float(l[1].strip()) * u.deg
+                    else:
+                        pass
```
```suggestion
```

eerovaher commented on astroquery/solarsystem/pds/tests/test_pds.py:
```diff
@@ -0,0 +1,219 @@
+    q = pds.RingNode()
+    # need _last_query to be defined
+    q._last_query = AstroQuery('GET', 'http://dummy')
+    with pytest.raises(ValueError):
```
It's good to test the exception message because the expected message also conveys the purpose of the test.

eerovaher commented on astroquery/solarsystem/pds/tests/test_pds.py:
```diff
@@ -0,0 +1,219 @@
+    mab = bodytable[bodytable.loc_indices["Mab"]]
+    assert mab["NAIF ID"] == 726
+    assert mab["Body"] == "Mab"
+    assert np.allclose(
+        [42.011201, 15.801323, 5.368, 0.623, 223.976, 55.906, 223.969, 56.013, 0.10932, 3098.514, ],
+        [mab["RA (deg)"].to(u.deg).value, mab["Dec (deg)"].to(u.deg).value, mab["dRA"].to(u.arcsec).value, mab["dDec"].to(u.arcsec).value, mab["sub_obs_lon"].to(u.deg).value, mab["sub_obs_lat"].to(u.deg).value, mab["sub_sun_lon"].to(u.deg).value, mab["sub_sun_lat"].to(u.deg).value, mab["phase"].to(u.deg).value, mab["distance"].to(u.km * 1e6).value, ],
+        rtol=1e-2,
+    )
```
I'd recommend using a separate `assert` for each column of the table. If this test should fail then with the way it is written currently written it would be quite difficult to figure out which column would be causing the problem.

eerovaher commented on astroquery/solarsystem/pds/tests/test_pds.py:
```diff
@@ -0,0 +1,219 @@
+        rtol=1e-2,
+    )
+
+    # check ringtable is None
```
```suggestion
```

eerovaher commented on astroquery/solarsystem/pds/tests/test_pds_remote.py:
```diff
@@ -0,0 +1,38 @@
+
+
+@pytest.mark.remote_data
+class TestRingNodeClass:
```
What's the purpose of the class?

eerovaher commented on .gitignore:
```diff
@@ -20,6 +20,7 @@ MANIFEST
 pip-wheel-metadata
 .hypothesis
 doctests.py
+coverage.xml
```
Why does `.gitignore` have to be updated?

bsipocz replied:
This is an OK addition

eerovaher reviewed (commented):
I don't claim to have done a very thorough review, but I did spot a few things worth pointing out.

bsipocz commented on astroquery/solarsystem/pds/core.py:
```diff
@@ -0,0 +1,427 @@
+
+        Parameters
+        ----------
+        planet : str, required. one of Mars, Jupiter, Saturn, Uranus, Neptune, or Pluto
```
I would just add a line break here. We do use `str` elsewhere and list the possible inputs in the detailed description.

bsipocz commented on astroquery/solarsystem/pds/core.py:
```diff
@@ -0,0 +1,427 @@
+        ringtable : `astropy.QTable`
+        """
+        try:
+            self._last_query.remove_cache_file(self.cache_location)
```
Why do we need to remove the cache?

bsipocz commented:
OK, so I think we kept this open long enough. I have addressed the last set of comments, but as there has been multiple rounds of many suggestions I could easily missed a few. Anyway, tests all pass and the module is functional, so any further improvements can be done in a follow-up PR.

bsipocz commented:
Thanks @emolter!
